### DISCUSSION TITLE
Swiping away alarm notifications doesn't always silence the alarm on Pixel 6 Pro

### DISCUSSION THREAD

**Discussion by thedewi** [Navid200/xDrip#135] [Alerts]
When an alert sounds, swiping it away in the system notification list *sometimes* does not silence the alarm tone, and it continues to play for a long time.

The only way I've found to stop the sound is to reduce the phone volume right down, or turn the phone off. Sometimes after a few minutes I can bring the audio volume back up and it has stopped.

xDrip+ gives every indication that the alarm has been dismissed, and does not mention any current alert. For example, disabling alerts gives no mention of any current alert, and will happily disable future alerts. But the audio of the alert keeps playing.

Putting the phone on silent does not stop the audio.

Force-stopping xDrip+ does not stop the audio.

The only way I've found to stop the alert sound is to turn the phone off.

Version: 49817a5-2021.10.04

**Comment by Navid200** [comment]
Would you please show the following page?

Settings -> Alarms and Alerts -> Glucose Alerts Settings

**Reply by thedewi** [reply]
![Screenshot_20220112-151702](https://user-images.githubusercontent.com/4726656/149068313-2f17e3d5-a7d0-44d6-af03-83f084476eb1.png)

**Comment by thedewi** [comment]
It seems to happen a lot more often when I'm using a Bluetooth headset, especially if it's "out of focus" due to something like a phone call - the alarm tones seem to queue up in the background, so when I come off the call I might hear 5+ alarms simultaneously, looping in perpetuity, and I have to restart the phone.

I am trying enabling "Notification Channels" to see if it behaves any differently.

**Comment by Navid200** [comment]
Do you have force speaker enabled?
https://github.com/Navid200/xDrip/wiki/Alert-raising-the-music-volume

**Reply by thedewi** [reply]
No, it only plays on speaker if no headset is connected. I will experiment with that option next.

**Reply by Navid200** [reply]
But, that wasn't the question.
Some xDrip items don't have perfect description because of limited space.
Please enable force speaker as a first step.
Please don't enable notification channels or anything else.  There is no need for trial and error.

**Reply by Navid200** [reply]
Also, how exactly do you swipe away the notification?
It only works if you swipe right.

**Reply by Navid200** [reply]
Let me change that.  I can guarantee that it works swiping right.
I cannot guarantee it would work swiping left.

**Reply by thedewi** [reply]
My answer to the question about "Force Speaker" was "no"; I was just providing additional context.

I have now enabled "Force Speaker" on my level alerts, but I rarely use those alerts - I mostly use "forecast low", "persistent high" and "missed reading", and I cannot find a "Force Speaker" setting for those.

I will try swiping right, but I don't think the problem is about clearing the alert, because all the other behaviours associated with clearing the alert are working fine - the main view and the snooze menus correctly update to show no current alert. It is only the media playback that continues, and only on the Pixel 6 Pro, and usually when a Bluetooth headset is paired.

**Reply by Navid200** [reply]
> I will try swiping right, but I don't think the problem is about clearing the alert, 

But, your first post says:
"swiping it away in the system notification list sometimes does not silence the alarm tone"

So, when you swipe away the alert, does it snooze the alert or not?

**Reply by thedewi** [reply]
It is the tone that continues, not any other aspects of the alert.

**Reply by thedewi** [reply]
So yes I would say the alert itself is snoozed.

**Reply by Navid200** [reply]
If the alert tone continues, why do you say the alert itself is snoozed?
I want to understand what's going on.  I am having a hard time.
Can you put yourself in my shoes?
Can you explain what's going on to someone who has not experienced it?

**Reply by thedewi** [reply]
Update:
Just noticed I got something wrong with my observations. A "forecast low" doesn't count as an "alert" in the traditional way. So the snooze menu shows "No active alert exists" even when a forecast low is present and not dismissed. And the red text on the main screen that says "Low predicted in 18 mins" is unaffected by any swiping I do.

So I don't really know whether xDrip+ is registering my "swipe" properly.

But there is still one other clue that the problem is audio-related: my broken unstoppable alarms *loop*, and normally these alerts do not loop at all. I think something is going wrong when the audio playback *starts*.



Original comment below:
---


Every non-audio aspect of alerts is behaving normally. Sometimes I swipe them away, and sometimes they self-resolve. In either case, the red text at the top of the main screen disappears correctly, and the indicator in the snooze menu changes correctly, to "No active alert exists".

That is supporting evidence that xDrip+ is successfully responding to stimuli that snooze or resolve alerts, and is successfully managing alert states *except* for audio playback.

The problem I am observing *only* involves audio: playback sometimes does not stop. Because it happened as soon as I got a Pixel 6 Pro, and because it happens more often with Bluetooth, and far more often when audio playback was "out of focus" due to another subsystem playing audio, I see this as evidence that something about the Android audio subsystem has changed, possibly relating to "audio focus" management, as described in the Android documentation here: https://android-doc.github.io/guide/topics/media/mediaplayer.html#audiofocus

I am not familiar with xDrip+ code or Android development, but I am wondering if:

- Perhaps a new Android audio state that xDrip+ does not expect is causing it to lose its handle to a `MediaPlayer` object?
- Perhaps when xDrip+ playback is out of focus, the "stop playback" commands are not being received by Android?

**Reply by Navid200** [reply]
Awesome!
I had no idea anyone had experienced this.

Do you mind moving this to the main repository?
The discussions in the main repository have just been opened for the first time a few days ago.
More developers look at those.
It may seem like a lot of work to move this.
But, I suggest just opening a new discussions there (https://github.com/NightscoutFoundation/xDrip/discussions) and just copy and pasting your posts that are important.

If you are agreeable to doing this, please make sure to copy this last post I am replying to for sure.
You can copy all your posts.  You can only copy the first post and this post.
Or, you can copy this post and add to it anything that may be missing as just a single post.
That part of it is all up to you.

Unfortunately, I cannot transfer the discussion.

If you prefer to just leave it as is here, it's fine too.

**Reply by Navid200** [reply]
By the way, when I swipe right on the notification I get with a forecast low, it snoozes that alert as well.  The same with persistent high.

**Comment by thedewi** [comment]
I just noticed my Event Log did log some things about alerts today, so I'll try to collect some examples of "working" and "non-working" event sequences over the next little while.

**Comment by thedewi** [comment]
This is still happening - almost every time - with modern bluetooth headphones paired like Bose Noise Cancelling 700 or Jabra Elite 7 Pro.

I am too busy trying to get my phone to shut up to actually treat lows.

The "forecast low" alarm normally plays the tone just once - but with this bug, it loops forever until the phone is completely restarted.

The app is unusable at this point.

**Reply by Navid200** [reply]
These discussions are only kept for archival purposes.  Posting in them is not recommended.  
Please use these instead:  https://github.com/NightscoutFoundation/xDrip/discussions

**Reply by thedewi** [reply]
Sorry, I forgot you mentioned that. I'll transfer.

**Comment by thedewi** [comment]
Sometime in the next few weeks I will try to get a development environment up and running to see if I can reproduce this on a custom build and explore what is going wrong with the audio channel.
